### DISCUSSION TITLE
Make connection between return nullable value and output nullable parameter.

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/csharplang#3516] [Language Ideas]
Today you can specify parameter => return value direction:

```cs
[return: NotNullIfNotNull("output")] string? GetString(int x, out int? output)
```

It means if `output` is not null, then the return value is not null.
However, it is impossible to specify the other direction: if return value is not null, then output is not null.

I suggest adding another attribute for this: `NotNullIfReturnNotNull`

```cs
string? GetString(int x, [NotNullIfReturnNotNull] out int? output)
```

**Comment by 333fred** [comment]
2 things:

1. We haven't really encountered any scenarios that would need this. Can you give a concrete example?
2. This type of flow analysis is significantly more complicated than the other direction, as the first is an input flowing to an output. This is an output flowing to a related output that can only be checked at a later point in time.

**Comment by NN---** [comment]
I have encountered it in my code when starting using nullable .

I have a method which returns not null when it succeeds and sets some output parameter.
On failure it returns null and output parameter is null.

There some workarounds like making everything output and return Boolean.
Or return an object which aggregates both using either named type or tuple.

Any workaround means rewriting code which is not a good option right now since it involves too much changes.

**Comment by 333fred** [comment]
> Any workaround means rewriting code which is not a good option right now since it involves too much changes.

We understand that, which is why there are lots of escape hatches built into nrt. However, I'm just pointing out that after 7 months or so, this is the first request that I'm aware of for this type of attribute, and that nullable analysis as it's built today is not equipped to handle it.

**Comment by NN---** [comment]
Unfortunately I have been busy with writing features, tests and improving stability of application for these 7 months.

I should have been started porting to nullable much earlier and open this request since long time ago;)

**Comment by 333fred** [comment]
> I should have been started porting to nullable much earlier and open this request since long time ago;)

No worries. In terms of feasibility, though, this request compared to your other open requests is geotagging pictures vs determining whether it's of a bird: https://xkcd.com/1425/

**Comment by spydacarnage** [comment]
As your function either effectively returns two nulls or two not-nulls,
could it instead return either null or a single not-null tuple (and ignore
the out variable)?

**Comment by NN---** [comment]
As I said it is a workaround and requires code change.
I suspect that I have no choice, otherwise there are too many null assertions ( variable! ) in the code.
